server/src/dto/validators/validation.ts

```diff
@@ -0,0 +1,71 @@
+import { versions as resourceImportVersions } from '@dto/validators/resource_import/versions';
+import { ValidatorConstraint, ValidatorConstraintInterface, ValidationArguments } from 'class-validator';
+import Ajv from 'ajv';
+
+const jsonSchemas = {
+  resource_import: resourceImportVersions,
+};
+
+const compareVersions = (version1: string, version2: string) => {
+  const v1 = version1.split('.').map(Number);
+  const v2 = version2.split('.').map(Number);
+  for (let i = 0; i < v1.length; i++) {
+    if (v1[i] > v2[i]) {
+      return true;
+    }
+    if (v1[i] < v2[i]) {
+      return false;
+    }
+  }
+  return true;
+};
+
+const getSchemaBasedOnAppVersion = async (version: string, schemaVersions: Record<string, any>[]) => {
+  for (const schemaVersion of schemaVersions) {
+    if (compareVersions(version, schemaVersion.value)) {
+      return await schemaVersion.schema;
+    }
+  }
+};
+
+const jsonValidator = (schema: Record<string, any>, data: Record<string, any>) => {
+  const ajv = new Ajv({ allErrors: true });
+  const validate = ajv.compile(schema);
+  validate(data);
+  return validate.errors;
+};
+
+export const customValidator = async (
+  data: Record<string, any>,
+  schemaName: string,
+  version: string
+): Promise<Record<string, any>[]> => {
+  const schemaVersions = jsonSchemas[schemaName];
+  const schema = await getSchemaBasedOnAppVersion(version, schemaVersions);
+  if (!schema) return [];
+  const errors = jsonValidator(schema, data);
+  return errors || [];
+};
+
+@ValidatorConstraint({ name: 'jsonSchemaValidator', async: true })
+export class JsonSchemaValidator implements ValidatorConstraintInterface {
+  async validate(object: any, args: ValidationArguments): Promise<boolean> {
+    const schemaName = args.constraints[0];
+    const version = args.constraints[1];
+    const data = args.object;
+    const errors = await customValidator(data, schemaName, version);
+    args.constraints[2] = errors;
+    return errors.length === 0;
+  }
+
+  defaultMessage(args: ValidationArguments) {
+    const errors: any[] = args.constraints[2];
+    const errorMessage = errors.map((error) => {
+      return `Path: ${error.instancePath}\n Error:${error.message}\n`;
```

Comment:
Can we use the Logger service in here to log the specific errors in the server and then just return a generic error in this place instead like:  `File not recognized. Please ensure it matches the expected format.` ?

All the errors being shown as a toast in the front end might not be understandable for the end user. Until we have a better error-showing approach we can keep it generic for now.